subsys/bluetooth/services/dis.c

```diff
@@ -59,21 +95,145 @@ static ssize_t read_pnp_id(struct bt_conn *conn,
+			       BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+			       read_str, NULL, dis_hardware_revision),
+#endif
+#ifdef CONFIG_BT_GATT_DIS_SOFTWARE_REVISION
+	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_SOFTWARE_REVISION,
+			       BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+			       read_str, NULL, dis_software_revision),
+#endif
+
 };
 
 static struct bt_gatt_service dis_svc = BT_GATT_SERVICE(attrs);
 
+#if defined(CONFIG_BT_SETTINGS) && defined(CONFIG_BT_GATT_DIS_DYNAMIC)
+static int dis_set(int argc, char **argv, size_t len_rd, settings_read_fn read,
+		   void *store)
+{
+	int len;
+
+	if (!strcmp(argv[0], "manuf")) {
+		len = read(store, &dis_manuf,
+			   sizeof(dis_manuf) - 1);
+		if (len < 0) {
+			BT_ERR("Failed to read manufacturer from storage"
+				       " (err %d)", len);
+		} else {
+			dis_manuf[len] = '\0';
+
+			BT_DBG("Manufacturer set to %s", dis_manuf);
+		}
+		return 0;
+	}
+	if (!strcmp(argv[0], "model")) {
+		len = read(store, &dis_model,
+			   sizeof(dis_model) - 1);
+		if (len < 0) {
+			BT_ERR("Failed to read model from storage"
+				       " (err %d)", len);
+		} else {
+			dis_model[len] = '\0';
+
+			BT_DBG("Model set to %s", dis_model);
+		}
+		return 0;
+	}
+#ifdef CONFIG_BT_GATT_DIS_SERIAL_NUMBER
+	if (!strcmp(argv[0], "serial")) {
+		len = read(store, &dis_serial_number,
+			   sizeof(dis_serial_number) - 1);
+		if (len < 0) {
+			BT_ERR("Failed to read serial number from storage"
+				       " (err %d)", len);
+		} else {
+			dis_serial_number[len] = '\0';
+
+			BT_DBG("Serial number set to %s",
+			       dis_serial_number);
+		}
+		return 0;
+	}
+#endif
+#ifdef CONFIG_BT_GATT_DIS_FIRMWARE_REVISION
+	if (!strcmp(argv[0], "fw")) {
+		len = read(store, &dis_firmware_revision,
+			   sizeof(dis_firmware_revision) - 1);
+		if (len < 0) {
+			BT_ERR("Failed to read firmware revision from storage"
+				       " (err %d)", len);
+		} else {
+			dis_firmware_revision[len] = '\0';
+
+			BT_DBG("Firmware revision set to %s",
+			       dis_firmware_revision);
+		}
+		return 0;
+	}
+#endif
+#ifdef CONFIG_BT_GATT_DIS_HARDWARE_REVISION
+	if (!strcmp(argv[0], "hw")) {
+		len = read(store, &dis_hardware_revision,
+			   sizeof(dis_hardware_revision) - 1);
+		if (len < 0) {
+			BT_ERR("Failed to read hardware revision from storage"
+				       " (err %d)", len);
+		} else {
+			dis_hardware_revision[len] = '\0';
+
+			BT_DBG("Hardware revision set to %s",
+			       dis_hardware_revision);
+		}
+		return 0;
+	}
+#endif
+#ifdef CONFIG_BT_GATT_DIS_SOFTWARE_REVISION
+	if (!strcmp(argv[0], "sw")) {
+		len = read(store, &dis_software_revision,
+			   sizeof(dis_software_revision) - 1);
+		if (len < 0) {
+			BT_ERR("Failed to read software revision from storage"
+				       " (err %d)", len);
+		} else {
+			dis_software_revision[len] = '\0';
+
+			BT_DBG("Software revision set to %s",
+			       dis_software_revision);
+		}
+		return 0;
+	}
+#endif
+}
+
+BT_SETTINGS_DEFINE(dis, dis_set, NULL, NULL);
```

Comment:
So this is loading some setting saved on flash but there is no code to store these so I wonder how it gets there in the first place? Furthermore, we do have STR options to set these which gets overwritten if set in flash, Id say we might not need these settings if that can be configured via Kconfig that is probably much easier to setup then have it somewhat stored manually just to be loaded. The general guidance for using the settings subsystem is for values that do change at runtime and need to be persisted, if it is just hardcoded as config option that should not need a setting even if there is a new firmware to be flashed that can just change the config option value.

Reply:
In the system I am working on these values can be changed after final compilation as the device could be sold under different brands with each their own versioning scheme. Thus these values will be written in the settings during production and read from setting during boot. so kconfig is not an option.

Reply:
> In the system I am working on these values can be changed after final compilation as the device could be sold under different brands with each their own versioning scheme. Thus these values will be written in the settings during production and read from setting during boot. so kconfig is not an option.

So the idea is that it always compiled with the same value which always gets overwritten by what is stored in the settings? Im not sure if there is a practical difference in compiling with a different prj.conf though versus fabricating the settings in flash, afaik for doing the latter there is no tool yet, or there is something on the works? Either way with this design we end up having 3 copies of the same string, one for config option, another stored as setting and yet another for storing the either one of the above, so I can see companies actually favoring rebuilding with a different prj.conf rather than fabricating the flash with settings like you are suggesting. Another question about this design is if by trying to fabricate the flash wouldn't this wipe out existing settings? Afaik to do this properly on an existing image one would have to open the flash fs, read all the stored settings (keys, CCC, etc) and then reconstruct the fs, with the config option alone I beleave this would not be necessary because that doesn't change the flash where the settings are stored.

Reply:
The solution will not prevent compiling with different prj.conf files and using this way of setting up the system but will just provide an alternative way of controlling this for those who needs this. The default one will be the way it is working today.
I do not know of any tools providing being in the works, but e.g. blue address, IRK and device name will be read during boot. They will have to come from somewhere - don't know how this is done.
BTW. In my system the flash(settings) are generated much later in the flow than the compilation of zephyr as it is part of a larger setup.

Reply:
Right, though it is still unclear to me how DFU would work with these values stored as settings, or you are saying that we can do partial writes that just overwrites some blocks? Perhaps there should be another Kconfig just to enable the settings loading, that way we only need to enable the extra storage if the settings loading is enabled otherwise you can assign the config options directly as user_data of the attributes.

Reply:
I changed it so now, when settings are not used it will work as previously using the kconfig.

My knowledge about the DFU is limited, as it is not used in the system I work on.